ard_controller/ard_controller.ino

```diff
@@ -15,45 +25,274 @@ struct JointAngleStruct {
+          preamble_counter--;
+          if (preamble_counter == 0) {
+            preamble_counter = PREAMBLE_LENGTH;
+            sr_state = READ_L_HIP;
+          }
+        }
+        else {
+          preamble_counter = PREAMBLE_LENGTH;
+        }
+        break;
+      case READ_L_HIP:
+        // this is different than other cases as it needs to catch a request for sensor data
+        if (data_byte_counter == 1 && temporary_packet_data.left_hip == received_data) {
+          temporary_packet_data.data_request = true;
+          temporary_packet_data.left_hip += received_data;
+          calculated_checksum += received_data;
+          data_byte_counter = DATA_BYTE_LENGTH;
+          sr_state = READ_CHECKSUM;
+        }
+        else {
+          temporary_packet_data.left_hip += received_data;
+          calculated_checksum += received_data;
+          data_byte_counter--;
+          if (data_byte_counter == 0) {
+            data_byte_counter = DATA_BYTE_LENGTH;
+            sr_state = READ_L_KNEE;
+          }
+        }
+        break;
+      case READ_L_KNEE:
+        temporary_packet_data.left_knee += received_data;
+        calculated_checksum += received_data;
+        data_byte_counter--;
+        if (data_byte_counter == 0) {
+          data_byte_counter = DATA_BYTE_LENGTH;
+          sr_state = READ_R_HIP;
+        }
+        break;
+      case READ_R_HIP:
+        temporary_packet_data.right_hip += received_data;
+        calculated_checksum += received_data;
+        data_byte_counter--;
+        if (data_byte_counter == 0) {
+          data_byte_counter = DATA_BYTE_LENGTH;
+          sr_state = READ_R_KNEE;
+        }
+        break;
+      case READ_R_KNEE:
+        temporary_packet_data.right_knee += received_data;
+        calculated_checksum += received_data;
+        data_byte_counter--;
+        if (data_byte_counter == 0) {
+          data_byte_counter = DATA_BYTE_LENGTH;
+          sr_state = READ_L_SHOULDER;
+        }
+        break;
+      case READ_L_SHOULDER:
+        temporary_packet_data.left_shoulder += received_data;
+        calculated_checksum += received_data;
+        data_byte_counter--;
+        if (data_byte_counter == 0) {
+          data_byte_counter = DATA_BYTE_LENGTH;
+          sr_state = READ_L_ELBOW;
+        }
+        break;
+      case READ_L_ELBOW:
+        temporary_packet_data.left_elbow += received_data;
+        calculated_checksum += received_data;
+        data_byte_counter--;
+        if (data_byte_counter == 0) {
+          data_byte_counter = DATA_BYTE_LENGTH;
+          sr_state = READ_R_SHOULDER;
+        }
+        break;
+      case READ_R_SHOULDER:
+        temporary_packet_data.right_shoulder += received_data;
+        calculated_checksum += received_data;
+        data_byte_counter--;
+        if (data_byte_counter == 0) {
+          data_byte_counter = DATA_BYTE_LENGTH;
+          sr_state = READ_R_ELBOW;
+        }
+        break;
+      case READ_R_ELBOW:
```

Comment:
Will we ever have a situation when we are asking for a specific sensor reading and not just all of them at once?

Reply:
There might be, but the primary purpose of the service is for the tracker ROS node to keep track of all joints and IMU angles to make sure things are on the right track and to determine when to send the next setpoint. Also, sending all sensor data versus data on specific sensors has a negligible difference in time cost, likely in the order of microseconds or even nanoseconds. Also also, adding the functionality to request specific sensors instead of all of them at once would introduce significant complexity to the state machine to read the serial input since now there would have to be bytes to identify what sensor the proceeding bytes represent instead of relying on the values being sent in a specific order.

Reply:
makes sense! haha that's def not how the current json pipeline works, but that's why we're developing this! :)